File: customer/src/main/kotlin/it/beaesthetic/wallet/domain/Wallet.kt
package it.beaesthetic.wallet.domain

import java.time.Instant
import java.util.*

@JvmInline value class WalletId(val value: String)

data class Wallet(
    val id: String,
    val owner: String,
    val availableAmount: Money,
    val spentAmount: Money,
    val operations: List<WalletEvent>,
    val giftCards: List<GiftCard>,
    val createdAt: Instant = Instant.now(),
    val updatedAt: Instant = Instant.now(),
    private val processorPolicy: GiftCardProcessorPolicy,
) {

    companion object {
        fun ofCustomer(customer: String): Wallet =
            Wallet(
                id = UUID.randomUUID().toString(),
                owner = customer,
                availableAmount = Money.Zero,
                spentAmount = Money.Zero,
                operations = listOf(),
                giftCards = listOf(),
                processorPolicy = GiftCardProcessorPolicy(),
            )
    }

    fun charge(amount: Money, now: Instant = Instant.now()): Wallet {
        require(amount > Money.Zero) { "Amount must be positive" }
        require(amount <= availableAmount) {
            "Cannot redeem $amount cause exceed maximum amount of $availableAmount"
        }
        val updatedWallet = removeExpiredGiftCard(now)
        val (residual, giftCards) = processorPolicy.charge(updatedWallet.giftCards, amount, now)
        require(residual == Money.Zero) { "Cannot redeem $amount from gift cards" }
        return copy(
            availableAmount = updatedWallet.availableAmount - amount,
            operations = listOf(MoneyCharge(now, amount)) + operations,
            spentAmount = updatedWallet.spentAmount + amount,
            giftCards = giftCards,
        )
    }

    fun creditMoney(amount: Money, now: Instant = Instant.now()): Wallet {
        return copy(
            availableAmount = availableAmount + amount,
            operations = listOf(MoneyCredited(now, amount)) + operations,
        )
    }

    fun creditGiftCard(giftCard: GiftCard, now: Instant = Instant.now()): Wallet {
        require(giftCard.owner == owner) { "Gift card owner is not the same as the owner" }
        require(giftCards.none { it.id == giftCard.id }) { "Gift card already redeemed" }
        require(!giftCard.isExpiredAt(now)) { "Expired gift card cannot be redeemed" }
        return copy(
            availableAmount = availableAmount + giftCard.availableAmount,
            operations =
                listOf(
                    GiftCardMoneyCredited(
                        at = now,
                        giftCardId = giftCard.id,
                        expiresAt = giftCard.expiresAt,
                        amount = giftCard.availableAmount,
                    )
                ) + operations,
            giftCards = listOf(giftCard) + giftCards,
        )
    }

    fun expireGiftCard(giftCard: GiftCard, now: Instant = Instant.now()): Wallet {
        val expiredGiftCard = giftCards.find { it.id == giftCard.id }
        if (expiredGiftCard != null) {
            return copy(
                availableAmount = availableAmount - expiredGiftCard.availableAmount,
                operations =
                    listOf(GiftCardMoneyExpired(at = now, giftCardId = giftCard.id)) + operations,
                giftCards = giftCards - expiredGiftCard,
            )
        }
        return this
    }

    private fun removeExpiredGiftCard(now: Instant = Instant.now()): Wallet {
        return giftCards
            .filter { it.isExpiredAt(now) }
            .fold(this) { wallet, card -> wallet.expireGiftCard(card) }
    }
}

File: customer/internal/domain/wallet/wallet.go
package wallet

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

const giftCardDuration = 365 * 24 * time.Hour

type Wallet struct {
	ID              string
	Owner           string
	AvailableAmount float64
	Spent           float64
	Operations      []Operation
	GiftCards       []GiftCard
	CreatedAt       time.Time
	UpdatedAt       time.Time
}

type GiftCard struct {
	ID              string    `json:"id"`
	Owner           string    `json:"owner"`
	AvailableAmount float64   `json:"availableAmount"`
	CreatedAt       time.Time `json:"createdAt"`
	ExpiresAt       time.Time `json:"expiresAt"`
	AmountSpent     float64   `json:"amountSpent"`
}

type Operation struct {
	Type       string    `json:"type"`
	Amount     float64   `json:"amount,omitempty"`
	At         time.Time `json:"at"`
	GiftCardID string    `json:"giftCardId,omitempty"`
	ExpireAt   time.Time `json:"expireAt,omitempty"`
}

func New(owner string, now time.Time) Wallet {
	return Wallet{ID: uuid.NewString(), Owner: owner, CreatedAt: now.UTC(), UpdatedAt: now.UTC()}
}

func (w Wallet) CreditGiftCard(amount float64, now time.Time) (Wallet, error) {
	if amount <= 0 {
		return Wallet{}, fmt.Errorf("amount must be positive")
	}
	card := GiftCard{ID: uuid.NewString(), Owner: w.Owner, AvailableAmount: amount, CreatedAt: now.UTC(), ExpiresAt: now.Add(giftCardDuration).UTC()}
	w.AvailableAmount += amount
	w.GiftCards = append([]GiftCard{card}, w.GiftCards...)
	w.Operations = append([]Operation{{Type: "giftCardMoneyCredited", Amount: amount, At: now.UTC(), GiftCardID: card.ID, ExpireAt: card.ExpiresAt}}, w.Operations...)
	w.UpdatedAt = now.UTC()
	return w, nil
}

func (w Wallet) Charge(amount float64, now time.Time) (Wallet, error) {
	if amount <= 0 {
		return Wallet{}, fmt.Errorf("amount must be positive")
	}
	w = w.removeExpiredGiftCards(now)
	if amount > w.AvailableAmount {
		return Wallet{}, fmt.Errorf("cannot redeem %.2f cause exceed maximum amount of %.2f", amount, w.AvailableAmount)
	}
	remaining := amount
	for idx := range w.GiftCards {
		if remaining <= 0 {
			break
		}
		card := &w.GiftCards[idx]
		if card.ExpiresAt.Before(now) {
			continue
		}
		charge := min(card.AvailableAmount, remaining)
		card.AvailableAmount -= charge
		card.AmountSpent += charge
		remaining -= charge
	}
	w.AvailableAmount -= amount
	w.Spent += amount
	w.Operations = append([]Operation{{Type: "moneyCharged", Amount: amount, At: now.UTC()}}, w.Operations...)
	w.UpdatedAt = now.UTC()
	return w, nil
}

func (w Wallet) removeExpiredGiftCards(now time.Time) Wallet {
	active := make([]GiftCard, 0, len(w.GiftCards))
	for _, card := range w.GiftCards {
		if card.ExpiresAt.Before(now) && card.AvailableAmount > 0 {
			w.AvailableAmount -= card.AvailableAmount
			w.Operations = append([]Operation{{Type: "giftCardMoneyExpired", Amount: card.AvailableAmount, At: now.UTC(), GiftCardID: card.ID}}, w.Operations...)
			continue
		}
		active = append(active, card)
	}
	w.GiftCards = active
	return w
}
